torch/nn/quantizable/modules/rnn.py

```diff
@@ -0,0 +1,294 @@
+                                     self.bias, batch_first=False,
+                                     bidirectional=self.bidirectional))
+        self.layers = torch.nn.ModuleList(layers)
+
+    def forward(self, x, hidden=None):
+        if self.batch_first:
+            x = x.transpose(0, 1)
+
+        max_batch_size = x.size(1)
+        num_directions = 2 if self.bidirectional else 1
+        if hidden is None:
+            zeros = torch.zeros(num_directions, max_batch_size,
+                                self.hidden_size, dtype=torch.float,
+                                device=x.device)
+            zeros.squeeze_(0)
+            if x.is_quantized:
+                zeros = torch.quantize_per_tensor(zeros, scale=1.0,
+                                                  zero_point=0, dtype=x.dtype)
+            hidden = [(zeros, zeros) for _ in range(self.num_layers)]
+        elif isinstance(hidden[0], Tensor):
+            hx = hidden[0].reshape(self.num_layers, num_directions,
+                                   max_batch_size, self.hidden_size).unbind(0)
+            cx = hidden[1].reshape(self.num_layers, num_directions,
+                                   max_batch_size, self.hidden_size).unbind(0)
+            hidden = []
+            for idx in range(self.num_layers):
+                hidden.append((hx[idx].squeeze_(0), cx[idx].squeeze_(0)))
+
+        for idx in range(self.num_layers):
+            x, hidden[idx] = self.layers[idx](x, hidden[idx])
+
+        hx_list = []
+        cx_list = []
+        for idx in range(self.num_layers):
+            hx_list.append(hidden[idx][0])
+            cx_list.append(hidden[idx][1])
+        hx_tensor = torch.stack(hx_list)
+        cx_tensor = torch.stack(cx_list)
+
+        # We are creating another dimension for bidirectional case
+        # need to collapse it
+        hx_tensor = hx_tensor.reshape(-1, *hx_tensor.shape[-2:])
+        cx_tensor = cx_tensor.reshape(-1, *cx_tensor.shape[-2:])
+
+        if self.batch_first:
+            x = x.transpose(0, 1)
+
+        return x, (hx_tensor, cx_tensor)
+
+    def _get_name(self):
+        return 'QuantizableLSTM'
+
+    @classmethod
+    def from_float(cls, other, qconfig=None):
+        assert isinstance(other, cls._FLOAT_MODULE)
+        assert (hasattr(other, 'qconfig') or qconfig)
+        observed = cls(other.input_size, other.hidden_size, other.num_layers,
+                       other.bias, other.batch_first, other.dropout,
+                       other.bidirectional)
+        observed.qconfig = getattr(other, 'qconfig', qconfig)
+        for idx in range(other.num_layers):
+            observed.layers[idx] = _LSTMLayer.from_float(other, idx, qconfig,
+                                                         batch_first=False)
+        observed.eval()
+        observed = torch.quantization.prepare(observed, inplace=True)
+        return observed
+
+    def from_observed(self, other):
```

Comment:
I see the calls to this in the custom module code, but this PR seems like the first non-test use case.  Is that expected?  Should this be an interface with a docblock somewhere?

If this is the only callsite, should we even have this API at all?  Can we just inline the `torch.quantization.convert` call in the custom module flow instead, and have one less API surface to worry about?  cc @jerryzh168

Reply:
We had a minor discussion with @jerry39213gh about exactly that -- the reason we want to have the `from_observed` is because the conversion process is not necessarily producing structurally equivalent module. For example, if we decide to rewrite the LSTM in C++, and have a module that just calls it in the `forward`, we could. In such a case, the `from_float` would convert the FP module into the "quantizable" version that will be calibrated/observed, and later converted using the `from_observed`. In such (hypothetical) scenario, the `from_observed` would not call the `convert`, but explicitly remove the internal linear layers and replace all of them with a single call to an `aten` kernel.

As for the docs: This will be documented the same way the `from_float` is documented: Once the custom module is more or less stable, we will add a tutorial + a paragraph in the quantization documentation